ndb/src/google/cloud/ndb/model.py

```diff
@@ -1542,11 +1543,123 @@ def _get_for_dict(self, entity):
         return self._get_value(entity)
 
 
+def _validate_key(value, entity=None):
+    """Validate a key.
+
+    Args:
+        value (.Key): The key to be validated.
+        entity (Optional[Model]): The entity that the key is being validated
+            for.
+
+    Returns:
+        .Key: The passed in ``value``.
+
+    Raises:
+        .BadValueError: If ``value`` is not a :class:`.Key`.
+        KindError: If ``entity`` is specified, by the kind of the entity
+            doesn't match the kind of ``value``.
+    """
+    if not isinstance(value, Key):
+        raise exceptions.BadValueError("Expected Key, got {!r}".format(value))
+
+    if entity and entity.__class__ not in (Model, Expando):
```

Comment:
is x.__class__ preferable to type(x)?

Reply:
Also, I'm a little confused. Don't we expect `entity` to be an instance of `Model` or `Expando`?

Reply:
I'm not sure if `.__class__` is preferable to `type()`, though they had a difference with old-style classes:

```python
>>> import sys
>>> sys.version
'2.7.15 (default, May 14 2018, 21:07:29) \n[GCC 5.4.0 20160609]'
>>> class A:
...     pass
...
>>> class B(object):
...     pass
...
>>> a = A()
>>> b = B()
>>> type(a)
<type 'instance'>
>>> type(b)
<class '__main__.B'>
>>> a.__class__
<class __main__.A at 0x7f928bf372c0>
>>> b.__class__
<class '__main__.B'>
```

Reply:
Re: `Model` / `Expando`, that is for cases where the class is literally one of those, i.e. not a subclass. We expect users to have

```python
class Book(ndb.Model):
    author = ndb.StringProperty()
    pages = ndb.IntegerProperty()
```

so that `book.__class__ is Book`, not `Model`.

Reply:
Re: Model / Expando

Yep, duh.

Reply:
In Python3 `type(o) is o.__class__`, afaict. I think `type(o)` is preferred to `o.__class__` for the same reason that `len(seq)` is preferred over `seq.__len__()`, but not a big deal if you have a style preference.

Reply:
[Argument][1] for `type()` over `.__class__` 

```python
>>> import sys
>>> sys.version
'3.7.1 (default, Oct 21 2018, 22:21:13) \n[GCC 7.3.0]'
>>> class A:
...     pass
...
>>> class B(A):
...     __class__ = A
...
>>> a = A()
>>> a.__class__
<class '__main__.A'>
>>> type(a)
<class '__main__.A'>
>>>
>>> b = B()
>>> b.__class__
<class '__main__.A'>
>>> type(b)
<class '__main__.B'>
```

[1]: https://stackoverflow.com/a/10633356/1068170